Missing RegisterIndicator overloads

See `RegisterIndicatorRegressionAlgorithm`:

```
self.register_indicator(_symbol, indicator2, consolidator, lambda bar: self.set_selector_called(0) and bar)
```

MyPy output:

```
RegisterIndicatorRegressionAlgorithm.py_pd6njxkw:47: error: Argument 4 to "register_indicator" of "QCAlgorithm" has incompatible type "Callable[[IBaseData], Any | IBaseData]"; expected "Callable[[IBaseData], float]"  [arg-type]
```

The stubs only have the overloads that take the selector returning a `float` because the other ones (with selectors that return IBaseData) are generic and not added to the stubs.

--- a/QuantConnectStubsGenerator.Tests/GeneratorTests.cs
+++ b/QuantConnectStubsGenerator.Tests/GeneratorTests.cs
@@ -195,22 +195,23 @@ namespace QuantConnect.GenericMethodsTest
             var baseNameSpace = namespaces.Single(x => x.Name == "QuantConnect");
             var testNameSpace = namespaces.Single(x => x.Name == "QuantConnect.GenericMethodsTest");
 
-            var testClass = testNameSpace.GetClasses().Single();
-            Assert.AreEqual("TestClass", testClass.Type.Name);
+            // The generic-method workaround hoists two helper classes to the namespace level
+            // (issue #82). TestClass keeps only the property; it has no History methods.
+            var testClass = testNameSpace.GetClasses().Single(x => x.Type.Name == "TestClass");
+            Assert.AreEqual(0, testClass.Methods.Count);
+            Assert.AreEqual(0, testClass.InnerClasses.Count);
+            Assert.IsNotNull(testClass.Properties.SingleOrDefault(x => x.Name == "History"));
 
-            var testMethodCount = testClass.Methods.Count;
-            Assert.AreEqual(0, testMethodCount);
+            // The dispatch helper is non-generic and exposes __call__ + __getitem__.
+            var indexableClass = testNameSpace.GetClasses().Single(x => x.Type.Name == "_History");
+            Assert.AreEqual(0, indexableClass.Type.TypeParameters?.Count ?? 0);
+            Assert.IsNotNull(indexableClass.Methods.SingleOrDefault(x => x.Name == "__getitem__"));
+            Assert.IsNotNull(indexableClass.Methods.SingleOrDefault(x => x.Name == "__call__"));
 
-            var innerClass = testClass.InnerClasses.Single();
-
-            Assert.AreEqual(0, innerClass.Type.TypeParameters?.Count ?? 0);
-            Assert.IsNotNull(innerClass.Methods.SingleOrDefault(x => x.Name == "__getitem__"));
-            Assert.IsNotNull(innerClass.Methods.SingleOrDefault(x => x.Name == "__call__"));
-
-            var innerInnerClass = innerClass.InnerClasses.Single();
-
-            Assert.AreNotEqual(0, innerInnerClass.Type.TypeParameters?.Count ?? 0);
-            Assert.AreEqual(2, innerInnerClass.Methods.Count(x => x.Name == "__call__"));
+            // The typed helper is generic and holds the Sequence[T]-returning overloads.
+            var typedClass = testNameSpace.GetClasses().Single(x => x.Type.Name == "_TypedHistory");
+            Assert.AreNotEqual(0, typedClass.Type.TypeParameters?.Count ?? 0);
+            Assert.AreEqual(2, typedClass.Methods.Count(x => x.Name == "__call__"));
         }
 
         [Test]

--- a/QuantConnectStubsGenerator.Tests/MethodParserTests.cs
+++ b/QuantConnectStubsGenerator.Tests/MethodParserTests.cs
@@ -178,5 +178,95 @@ namespace QuantConnect.MethodParserTests
             var method4 = testClass.Methods.Single(x => x.Name == "TestMethod4");
             Assert.AreEqual(expectedConvertedType, method4.ReturnType);
         }
+
+        [Test]
+        public void KeepsCSharpOverloadWhenPythonPartialSharesSignature()
+        {
+            // Reproduces the QCAlgorithm.Link / BroadcastCommand regression:
+            // both partial files collapse to the same Python signature, and when the
+            // .Python.cs partial is visited first, PostProcessClass was deleting the
+            // survivor because the only entry in the HashSet had File=".Python.cs".
+            var testGenerator = new TestGenerator
+            {
+                Files = new()
+                {
+                    { "TestClass.Python.cs", @"
+using Python.Runtime;
+
+namespace QuantConnect.MethodParserTests
+{
+    public partial class TestClass
+    {
+        public string Link(PyObject command) { return null; }
+    }
+}" },
+                    { "TestClass.cs", @"
+namespace QuantConnect.MethodParserTests
+{
+    public partial class TestClass
+    {
+        public string Link(object command) { return null; }
+    }
+}" }
+                }
+            };
+
+            var result = testGenerator.GenerateModelsPublic();
+
+            var testNameSpace = result.GetNamespaces().Single(x => x.Name == "QuantConnect.MethodParserTests");
+            var testClass = testNameSpace.GetClasses().Single(x => x.Type.Name == "TestClass");
+
+            var linkMethod = testClass.Methods.SingleOrDefault(x => x.Name == "Link");
+            Assert.IsNotNull(linkMethod, "Link should be kept when the C# partial provides it alongside a PyObject overload");
+            Assert.IsFalse(linkMethod.File.EndsWith(".Python.cs"),
+                "The surviving Link should come from the non-.Python.cs partial so PostProcessClass does not remove it");
+            Assert.AreEqual(1, linkMethod.Parameters.Count);
+        }
+
+        [Test]
+        public void IndicatorBaseParametersAcceptPythonIndicator()
+        {
+            var testGenerator = new TestGenerator
+            {
+                Files = new()
+                {
+                    { "Test.cs", @"
+namespace QuantConnect.Indicators
+{
+    public class IndicatorDataPoint { }
+
+    public class IndicatorBase<T> { }
+
+    public class TestClass
+    {
+        public void TestMethod(string symbol, IndicatorBase<IndicatorDataPoint> indicator)
+        {
+        }
+    }
+}" }
+            }
+            };
+
+            var result = testGenerator.GenerateModelsPublic();
+
+            var ns = result.GetNamespaces().Single(x => x.Name == "QuantConnect.Indicators");
+            var testClass = ns.GetClasses().Single(x => x.Type.Name == "TestClass");
+
+            var method = testClass.Methods.Single(x => x.Name == "TestMethod");
+            var indicatorParam = method.Parameters[1];
+
+            // Should be Union[IndicatorBase[IndicatorDataPoint], PythonIndicator]
+            Assert.AreEqual("Union", indicatorParam.Type.Name);
+            Assert.AreEqual("typing", indicatorParam.Type.Namespace);
+            Assert.AreEqual(2, indicatorParam.Type.TypeParameters.Count);
+
+            var indicatorBaseType = indicatorParam.Type.TypeParameters[0];
+            Assert.AreEqual("IndicatorBase", indicatorBaseType.Name);
+            Assert.AreEqual("IndicatorDataPoint", indicatorBaseType.TypeParameters[0].Name);
+
+            var pythonIndicatorType = indicatorParam.Type.TypeParameters[1];
+            Assert.AreEqual("PythonIndicator", pythonIndicatorType.Name);
+            Assert.AreEqual("QuantConnect.Indicators", pythonIndicatorType.Namespace);
+        }
     }
 }

--- a/QuantConnectStubsGenerator.Tests/MethodParserTests.cs
+++ b/QuantConnectStubsGenerator.Tests/MethodParserTests.cs
@@ -184,8 +184,8 @@ namespace QuantConnect.MethodParserTests
         {
             // Reproduces the QCAlgorithm.Link / BroadcastCommand regression:
             // both partial files collapse to the same Python signature, and when the
-            // .Python.cs partial is visited first, PostProcessClass was deleting the
-            // survivor because the only entry in the HashSet had File=".Python.cs".
+            // PyObject-taking partial is visited first, PostProcessClass was deleting the
+            // survivor because the only entry in the HashSet was the PyObject variant.
             var testGenerator = new TestGenerator
             {
                 Files = new()
@@ -218,8 +218,8 @@ namespace QuantConnect.MethodParserTests
 
             var linkMethod = testClass.Methods.SingleOrDefault(x => x.Name == "Link");
             Assert.IsNotNull(linkMethod, "Link should be kept when the C# partial provides it alongside a PyObject overload");
-            Assert.IsFalse(linkMethod.File.EndsWith(".Python.cs"),
-                "The surviving Link should come from the non-.Python.cs partial so PostProcessClass does not remove it");
+            Assert.IsFalse(linkMethod.HasPyObjectParameter,
+                "The surviving Link should come from the C# partial so PostProcessClass does not remove it");
             Assert.AreEqual(1, linkMethod.Parameters.Count);
         }
 

--- a/QuantConnectStubsGenerator/Generator.cs
+++ b/QuantConnectStubsGenerator/Generator.cs
@@ -182,11 +182,17 @@ namespace QuantConnectStubsGenerator
                 .Select(project => $"{_leanPath}/{project}")
                 .ToList();
 
-            // Find all C# files in non-blacklisted projects in Lean
+            // Find all C# files in non-blacklisted projects in Lean. Normalize path separators
+            // once per file so the blacklist regex and prefix checks behave identically on
+            // Windows and Linux.
             var sourceFiles = Directory
                 .EnumerateFiles(_leanPath, "*.cs", SearchOption.AllDirectories)
-                .Where(file => !blacklistedRegex.Any(regex => regex.IsMatch(file)))
-                .Where(file => !blacklistedPrefixes.Any(file.Replace("\\", "/").StartsWith))
+                .Where(file =>
+                {
+                    var normalized = file.Replace("\\", "/");
+                    return !blacklistedRegex.Any(regex => regex.IsMatch(normalized))
+                        && !blacklistedPrefixes.Any(normalized.StartsWith);
+                })
                 .ToList();
 
             // Find all relevant C# files in the C# runtime
@@ -285,6 +291,19 @@ namespace QuantConnectStubsGenerator
 
             HandleGenericMethods(cls);
 
+            // Precompute non-.Python.cs signatures once to keep the .Python.cs filter O(N)
+            // instead of O(N²). The signature captures what Python.cs overloads need to match
+            // on to be considered redundant: same name + same (ordered) parameter types.
+            var csharpSignatures = cls.Methods
+                .Where(m => m.File != null && !m.File.EndsWith(".Python.cs"))
+                .Select(m => (m.Name, Params: m.Parameters.Select(p => p.Type).ToList()))
+                .ToList();
+
+            bool HasMatchingCSharpOverload(Method m) => csharpSignatures.Any(s =>
+                s.Name == m.Name
+                && s.Params.Count == m.Parameters.Count
+                && s.Params.Zip(m.Parameters, (t, p) => t.Equals(p.Type)).All(match => match));
+
             var pythonMethodsToRemove = cls.Methods
                 .Where(m => m.File != null && m.File.EndsWith(".Python.cs"))
                 // Python implementations of logging methods accept any parameter type
@@ -303,6 +322,12 @@ namespace QuantConnectStubsGenerator
                     return paramName != "type" && paramName != "dataType" && paramName != "T" &&
                         !m.Parameters.Any(p => p.Name == "handler" && p.Type == PythonType.Any);
                 })
+                // Only drop a .Python.cs method when a non-.Python.cs overload with the same
+                // signature already exists. Standalone Python-only overloads — e.g.
+                // Download(str, PyObject headers) whose C# counterpart takes
+                // List[KeyValuePair[str, str]] — stay put; removing them would lose the
+                // dict-friendly signature pythonnet exposes at runtime.
+                .Where(HasMatchingCSharpOverload)
                 .ToList();
 
             foreach (var pythonMethod in pythonMethodsToRemove)
@@ -314,6 +339,19 @@ namespace QuantConnectStubsGenerator
 
                 cls.Methods.Remove(pythonMethod);
             }
+
+            // Drop obsolete overloads that are pure back-compat shims — a non-obsolete same-name
+            // method with the same return type already covers the behavior. Keeping them forces
+            // Python subclass overrides to implement every signature, producing spurious
+            // "Signature incompatible with supertype" mypy errors. Obsolete overloads that
+            // differ in return type (e.g. Liquidate returning List[int] vs List[OrderTicket])
+            // are preserved because their behavior isn't captured by the other overloads.
+            var liveOverloadSignatures = cls.Methods
+                .Where(m => m.DeprecationReason == null)
+                .Select(m => (m.Name, m.ReturnType))
+                .ToHashSet();
+            cls.Methods.RemoveWhere(m => m.DeprecationReason != null
+                && liveOverloadSignatures.Contains((m.Name, m.ReturnType)));
         }
 
         private void MarkOverloads(Class cls)
@@ -457,13 +495,40 @@ namespace QuantConnectStubsGenerator
                 var indexer = new Method("__getitem__", newGenericClass.Type);
                 indexer.Parameters.Add(new Parameter("type", new PythonType("Type", "typing") { TypeParameters = [genericType] }));
                 newIndexableClass.Methods.Add(indexer);
-                foreach (var methods in cls.Methods.Where(x => x.Name == genericMethodName))
+                var allMethods = cls.Methods.Where(x => x.Name == genericMethodName).ToList();
+
+                // Collect first-parameter types covered by Python.cs overloads (e.g. Symbol,
+                // List[Symbol], List[str] from a Union[Symbol, List[Symbol], List[str]] tickers
+                // parameter). C# overloads whose first parameter overlaps with any of these
+                // are skipped so the Python.cs overloads — which return pandas.DataFrame — take
+                // precedence over the C# IEnumerable-returning counterparts.
+                static IEnumerable<PythonType> FlattenUnion(PythonType type)
                 {
-                    var targetClass = newIndexableClass;
-                    if (methods.IsGeneric)
+                    return type.Namespace == "typing" && type.Name == "Union"
+                        ? type.TypeParameters
+                        : new[] { type };
+                }
+                var pythonFirstParamTypes = allMethods
+                    .Where(x => !x.IsGeneric && x.File?.EndsWith(".Python.cs") == true && x.Parameters.Count > 0)
+                    .SelectMany(x => FlattenUnion(x.Parameters[0].Type))
+                    .Select(tp => tp.ToPythonString())
+                    .ToHashSet();
+
+                foreach (var methods in allMethods)
+                {
+                    // Skip C# overloads whose first parameter overlaps with a Python.cs overload
+                    // (e.g. History(Symbol, ...) and History(List<Symbol>, ...) are both overridden
+                    // by History(PyObject tickers, ...) whose tickers accepts Symbol/List[Symbol]/List[str]).
+                    if (!methods.IsGeneric
+                        && pythonFirstParamTypes.Count > 0
+                        && methods.File?.EndsWith(".Python.cs") != true
+                        && methods.Parameters.Count > 0
+                        && FlattenUnion(methods.Parameters[0].Type).Any(t => pythonFirstParamTypes.Contains(t.ToPythonString())))
                     {
-                        targetClass = newGenericClass;
+                        continue;
                     }
+
+                    var targetClass = methods.IsGeneric ? newGenericClass : newIndexableClass;
                     targetClass.Methods.Add(new Method("__call__", methods) { Overload = true });
                 }
                 var property = new Property(genericMethodName)

--- a/QuantConnectStubsGenerator/Generator.cs
+++ b/QuantConnectStubsGenerator/Generator.cs
@@ -111,7 +111,9 @@ namespace QuantConnectStubsGenerator
                     reflectionClass.Methods.Remove(badMethod);
                 }
 
-                foreach (var cls in ns.GetClasses())
+                // Snapshot via ToList so PostProcessClass can register additional helper classes
+                // (e.g. the hoisted _TypedHistory generic helper) without invalidating the iterator.
+                foreach (var cls in ns.GetClasses().ToList())
                 {
                     // Remove Python implementations for methods where there is both a Python as well as a C# implementation
                     // The parsed C# implementation is usually more useful for autocomplete
@@ -289,13 +291,14 @@ namespace QuantConnectStubsGenerator
 
             HandleSymbolGenericClass(cls, context);
 
-            HandleGenericMethods(cls);
+            HandleGenericMethods(cls, context);
 
-            // Precompute non-.Python.cs signatures once to keep the .Python.cs filter O(N)
-            // instead of O(N²). The signature captures what Python.cs overloads need to match
-            // on to be considered redundant: same name + same (ordered) parameter types.
+            // Precompute non-PyObject signatures once to keep the filter O(N) instead of O(N²).
+            // We classify "C# equivalent" by the absence of a PyObject parameter rather than by
+            // ".Python.cs" file suffix so that partial classes declaring PyObject overloads in
+            // plain .cs files (e.g. some model classes) are handled correctly.
             var csharpSignatures = cls.Methods
-                .Where(m => m.File != null && !m.File.EndsWith(".Python.cs"))
+                .Where(m => !m.HasPyObjectParameter)
                 .Select(m => (m.Name, Params: m.Parameters.Select(p => p.Type).ToList()))
                 .ToList();
 
@@ -305,7 +308,7 @@ namespace QuantConnectStubsGenerator
                 && s.Params.Zip(m.Parameters, (t, p) => t.Equals(p.Type)).All(match => match));
 
             var pythonMethodsToRemove = cls.Methods
-                .Where(m => m.File != null && m.File.EndsWith(".Python.cs"))
+                .Where(m => m.HasPyObjectParameter)
                 // Python implementations of logging methods accept any parameter type
                 // C# implementations only accept strings/numbers, so we cannot remove the Python implementations
                 .Where(m => !loggingMethods.Contains($"{cls.Type.Name}.{m.Name}"))
@@ -322,7 +325,7 @@ namespace QuantConnectStubsGenerator
                     return paramName != "type" && paramName != "dataType" && paramName != "T" &&
                         !m.Parameters.Any(p => p.Name == "handler" && p.Type == PythonType.Any);
                 })
-                // Only drop a .Python.cs method when a non-.Python.cs overload with the same
+                // Only drop a PyObject-taking method when a non-PyObject overload with the same
                 // signature already exists. Standalone Python-only overloads — e.g.
                 // Download(str, PyObject headers) whose C# counterpart takes
                 // List[KeyValuePair[str, str]] — stay put; removing them would lose the
@@ -471,37 +474,68 @@ namespace QuantConnectStubsGenerator
         }
 
         /// <summary>
-        /// Helper method to support 'self.history(...)' and 'self.history[XYZ](...)' use cases at the same time. Solution is of the format:
+        /// Helper method to support 'self.history(...)' and 'self.history[XYZ](...)' at the same time.
+        /// Generated layout (issue #82: both helpers hoisted to namespace level with private names
+        /// so they don't appear as QCAlgorithm inner classes in IDE autocomplete; explicit
+        /// __getitem__ overloads are emitted for common data types so the set is discoverable):
         ///
-        /// class MyClass :
-        ///     class History :
-        ///         class History(typing.Generic[T]):
-        ///              def __call__(self, ticker: str) -> list[QuantConnect.Data.Market.DataDictionary[T]]: ...
-        ///         def __call__(self, ticker: str) -> pandas.DataFrame: ...
-        ///         def __getitem__(self, type: typing.Type[T]) -> History[T]: ...
+        /// class _History:
+        ///     def __call__(self, ticker: str) -> pandas.DataFrame: ...
+        ///     def __getitem__(self, type: typing.Type[T]) -> _TypedHistory[T]: ...
+        ///
+        /// class _TypedHistory(typing.Generic[T]):
+        ///     def __call__(self, ticker: str) -> list[QuantConnect.Data.Market.DataDictionary[T]]: ...
+        ///
+        /// class QCAlgorithm:
         ///     @property
-        ///     def history(self) -> History: ...
+        ///     def history(self) -> _History: ...
         /// </summary>
-        private void HandleGenericMethods(Class cls)
+        private void HandleGenericMethods(Class cls, ParseContext context)
         {
             // For all generic methods we need to perform a workaround see https://github.com/QuantConnect/quantconnect-stubs-generator/issues/38
             var genericMethodNames = cls.Methods.Where(x => x.IsGeneric).Select(x => x.Name).ToHashSet();
             foreach (var genericMethodName in genericMethodNames)
             {
                 var genericType = cls.Methods.Where(x => x.IsGeneric && x.Name == genericMethodName).First().GenericType;
-                var genericClassType = new PythonType(genericMethodName) { TypeParameters = [genericType] };
-                var newGenericClass = new Class(genericClassType) { Summary = string.Empty };
-                var newIndexableClass = new Class(new PythonType(genericMethodName)) { Summary = string.Empty };
-                var indexer = new Method("__getitem__", newGenericClass.Type);
-                indexer.Parameters.Add(new Parameter("type", new PythonType("Type", "typing") { TypeParameters = [genericType] }));
-                newIndexableClass.Methods.Add(indexer);
-                var allMethods = cls.Methods.Where(x => x.Name == genericMethodName).ToList();
 
-                // Collect first-parameter types covered by Python.cs overloads (e.g. Symbol,
-                // List[Symbol], List[str] from a Union[Symbol, List[Symbol], List[str]] tickers
-                // parameter). C# overloads whose first parameter overlaps with any of these
-                // are skipped so the Python.cs overloads — which return pandas.DataFrame — take
-                // precedence over the C# IEnumerable-returning counterparts.
+                // Helper classes — both hoisted to namespace level with private names so they
+                // don't clutter QCAlgorithm's inner-class namespace in IDE autocomplete (issue #82).
+                var indexableClassName = $"_{genericMethodName}";
+                var typedClassName = $"_Typed{genericMethodName}";
+                var typedClassType = new PythonType(typedClassName, cls.Type.Namespace) { TypeParameters = [genericType] };
+                var newGenericClass = new Class(typedClassType) { Summary = string.Empty };
+                var newIndexableClass = new Class(new PythonType(indexableClassName, cls.Type.Namespace)) { Summary = string.Empty };
+
+                // PyObject wrapper overloads first so the DataFrame-returning variants take
+                // precedence in mypy/pyright overload resolution when an arg matches both a
+                // typed C# overload (e.g. History(Symbol, int, Resolution)) and the wrapper.
+                var allMethods = cls.Methods.Where(x => x.Name == genericMethodName)
+                    .OrderBy(x => x.HasPyObjectParameter ? 0 : 1)
+                    .ToList();
+
+                // Scoped widening: the PyObject History wrapper's `tickers` parameter is declared
+                // as PyObject (→ Any). Widen to the concrete types it accepts at runtime so Symbol
+                // and the other supported shapes match the DataFrame-returning overload before
+                // the typed C# IEnumerable-returning overloads get a chance. Applied here (not
+                // globally in MethodParser) to avoid affecting non-History methods that happen
+                // to take a parameter named "tickers". Widening is performed on the copies we
+                // push into the helper classes below, not on the originals in cls.Methods —
+                // mutating a Method already in a HashSet would desync its hash and break the
+                // later RemoveWhere that evicts the moved methods from the outer class.
+                var tickersUnion = PythonType.CreateUnion(
+                    new PythonType("Symbol", "QuantConnect"),
+                    new PythonType("str"),
+                    new PythonType("List", "typing") { TypeParameters = { new PythonType("Symbol", "QuantConnect") } },
+                    new PythonType("List", "typing") { TypeParameters = { new PythonType("str") } },
+                    new PythonType("Universe", "QuantConnect.Data.UniverseSelection"),
+                    new PythonType("Type", "typing")
+                );
+
+                // Collect first-parameter types covered by PyObject wrapper overloads (e.g.
+                // Symbol, List[Symbol], List[str] from the widened tickers union). C# overloads
+                // whose first parameter overlaps with any of these are skipped so the wrapper
+                // overloads — which return pandas.DataFrame — take precedence over the C#
+                // IEnumerable-returning counterparts.
                 static IEnumerable<PythonType> FlattenUnion(PythonType type)
                 {
                     return type.Namespace == "typing" && type.Name == "Union"
@@ -509,36 +543,67 @@ namespace QuantConnectStubsGenerator
                         : new[] { type };
                 }
                 var pythonFirstParamTypes = allMethods
-                    .Where(x => !x.IsGeneric && x.File?.EndsWith(".Python.cs") == true && x.Parameters.Count > 0)
+                    .Where(x => !x.IsGeneric && x.HasPyObjectParameter && x.Parameters.Count > 0)
                     .SelectMany(x => FlattenUnion(x.Parameters[0].Type))
                     .Select(tp => tp.ToPythonString())
                     .ToHashSet();
 
                 foreach (var methods in allMethods)
                 {
-                    // Skip C# overloads whose first parameter overlaps with a Python.cs overload
-                    // (e.g. History(Symbol, ...) and History(List<Symbol>, ...) are both overridden
-                    // by History(PyObject tickers, ...) whose tickers accepts Symbol/List[Symbol]/List[str]).
+                    // Skip C# overloads whose first parameter is fully covered by a wrapper
+                    // overload (e.g. History(IEnumerable<Symbol>, ...) is covered by
+                    // History(PyObject tickers, ...) where tickers accepts List[Symbol]).
+                    // If any flattened type is NOT in the wrapper's set (e.g. a Symbol
+                    // first parameter that the wrapper's tickers Union does not include),
+                    // keep the C# overload so its distinct return type is preserved.
                     if (!methods.IsGeneric
                         && pythonFirstParamTypes.Count > 0
-                        && methods.File?.EndsWith(".Python.cs") != true
+                        && !methods.HasPyObjectParameter
                         && methods.Parameters.Count > 0
-                        && FlattenUnion(methods.Parameters[0].Type).Any(t => pythonFirstParamTypes.Contains(t.ToPythonString())))
+                        && FlattenUnion(methods.Parameters[0].Type).All(t => pythonFirstParamTypes.Contains(t.ToPythonString())))
                     {
                         continue;
                     }
 
                     var targetClass = methods.IsGeneric ? newGenericClass : newIndexableClass;
-                    targetClass.Methods.Add(new Method("__call__", methods) { Overload = true });
+                    var callMethod = new Method("__call__", methods) { Overload = true };
+                    if (methods.HasPyObjectParameter)
+                    {
+                        // Replace the Parameter instance (don't mutate — the original is shared
+                        // with cls.Methods and mutating would desync its hash there).
+                        for (var i = 0; i < callMethod.Parameters.Count; i++)
+                        {
+                            var p = callMethod.Parameters[i];
+                            if (p.Name == "tickers" && p.Type == PythonType.Any)
+                            {
+                                callMethod.Parameters[i] = new Parameter(p.Name, tickersUnion) { VarArgs = p.VarArgs, Value = p.Value };
+                            }
+                        }
+                    }
+                    targetClass.Methods.Add(callMethod);
                 }
+
+                // __getitem__ — single generic Type[T] form. Per-type explicit overloads
+                // were tried for discoverability (issue #82) but caused mypy to pick the
+                // first overload (e.g. Type[TradeBar]) for any Type[X] argument it couldn't
+                // strictly disprove, breaking custom-data-type history requests. The generic
+                // form correctly infers T from the argument for both mypy and pyright.
+                var indexer = new Method("__getitem__", typedClassType);
+                indexer.Parameters.Add(new Parameter("type", new PythonType("Type", "typing") { TypeParameters = [genericType] }));
+                newIndexableClass.Methods.Add(indexer);
+
                 var property = new Property(genericMethodName)
                 {
                     Type = newIndexableClass.Type,
                     Class = cls,
                 };
                 cls.Properties.Add(property);
-                cls.InnerClasses.Add(newIndexableClass);
-                newIndexableClass.InnerClasses.Add(newGenericClass);
+
+                // Hoist both helper classes to namespace level (siblings of cls), not nested.
+                var ns = context.GetNamespaceByName(cls.Type.Namespace);
+                ns.RegisterClass(newIndexableClass);
+                ns.RegisterClass(newGenericClass);
+
                 // remove those we've moved around
                 cls.Methods.RemoveWhere(x => x.Name.Equals(genericMethodName, StringComparison.InvariantCultureIgnoreCase));
             }

--- a/QuantConnectStubsGenerator/Model/Method.cs
+++ b/QuantConnectStubsGenerator/Model/Method.cs
@@ -41,6 +41,14 @@ namespace QuantConnectStubsGenerator.Model
 
         public bool AvoidImplicitTypes { get; set; }
 
+        /// <summary>
+        /// True when at least one parameter was originally declared as PyObject in C#.
+        /// Used as a more reliable signal than the ".Python.cs" file suffix to identify
+        /// Python-equivalent overloads — some partial classes declare PyObject-taking
+        /// methods in plain ".cs" files alongside their C# counterparts.
+        /// </summary>
+        public bool HasPyObjectParameter { get; set; }
+
         public Method(string name, PythonType returnType)
         {
             Name = name;
@@ -58,6 +66,7 @@ namespace QuantConnectStubsGenerator.Model
             GenericType = other.GenericType;
             DeprecationReason = other.DeprecationReason;
             Documentation = other.Documentation;
+            HasPyObjectParameter = other.HasPyObjectParameter;
         }
 
         public bool Equals(Method other)

--- a/QuantConnectStubsGenerator/Parser/MethodParser.cs
+++ b/QuantConnectStubsGenerator/Parser/MethodParser.cs
@@ -241,7 +241,13 @@ namespace QuantConnectStubsGenerator.Parser
                 return null;
             }
 
-            var originalReturnType = returnType;
+            // pythonnet materializes IEnumerable<T> return values as indexable, sized
+            // collections, and user code routinely calls len()/[i]/.index() on them.
+            // Rewrite typing.Iterable[T] return types to typing.Sequence[T] so mypy accepts
+            // those valid runtime patterns while keeping the return type covariant — this
+            // matters when a subclass overrides the method with a narrower element type.
+            // Inheritance lists are untouched.
+            returnType = IterableReturnToSequence(returnType);
 
             var method = new Method(name, returnType)
             {
@@ -309,6 +315,20 @@ namespace QuantConnectStubsGenerator.Parser
 
             method.GenericType = genericType;
             method.AvoidImplicitTypes = avoidImplicitConversionTypes;
+
+            // When partial classes define the same signature in both a .Python.cs file and the
+            // plain .cs file (e.g. QCAlgorithm.Link(PyObject) vs QCAlgorithm.Link(object), both
+            // normalized to `command: Any`), HashSet dedupe is first-wins and ignores File.
+            // If the .Python.cs variant is visited first, PostProcessClass later deletes it
+            // as an override with no C# counterpart. Drop the .Python.cs entry so the .cs
+            // version can take its place.
+            if (!method.File.EndsWith(".Python.cs")
+                && _currentClass.Methods.TryGetValue(method, out var existing)
+                && existing.File.EndsWith(".Python.cs"))
+            {
+                _currentClass.Methods.Remove(existing);
+            }
+
             _currentClass.Methods.Add(method);
 
             ImprovePythonAccessorIfNecessary(method);
@@ -347,6 +367,16 @@ namespace QuantConnectStubsGenerator.Parser
                         new PythonType("PythonConsolidator", "QuantConnect.Python"), new PythonType("timedelta", "datetime"));
                 }
 
+                // IndicatorBase<IndicatorDataPoint> parameters also accept PythonIndicator (IndicatorBase<IBaseData>)
+                if (parameter.Type.Namespace == "QuantConnect.Indicators"
+                    && parameter.Type.Name == "IndicatorBase"
+                    && parameter.Type.TypeParameters.Count == 1
+                    && parameter.Type.TypeParameters[0].Name == "IndicatorDataPoint")
+                {
+                    parameter.Type = PythonType.CreateUnion(parameter.Type,
+                        new PythonType("PythonIndicator", "QuantConnect.Indicators"));
+                }
+
                 // datetime parameters also accept dates
                 if (parameter.Type.Namespace == "datetime" && parameter.Type.Name == "datetime")
                 {
@@ -360,6 +390,21 @@ namespace QuantConnectStubsGenerator.Parser
                 && (parameter.Name == "type" || parameter.Name == "dataType" || parameter.Name == "T"))
             {
                 parameter.Type = new PythonType("Type", "typing");
+            }
+
+            // PyObject tickers parameter: the Python.cs History wrapper accepts Symbol,
+            // List[Symbol], List[str], Universe, and Type (e.g. self.history(Fundamental, ...)),
+            // returning pandas.DataFrame in each case. Widen the typing so mypy picks this
+            // overload over the C# IEnumerable one.
+            if (parameter.Name == "tickers" && parameter.Type == PythonType.Any)
+            {
+                parameter.Type = PythonType.CreateUnion(
+                    new PythonType("Symbol", "QuantConnect"),
+                    new PythonType("List", "typing") { TypeParameters = { new PythonType("Symbol", "QuantConnect") } },
+                    new PythonType("List", "typing") { TypeParameters = { new PythonType("str") } },
+                    new PythonType("Universe", "QuantConnect.Data.UniverseSelection"),
+                    new PythonType("Type", "typing")
+                );
             }
 
             // System.Object parameters can accept anything
@@ -542,6 +587,31 @@ namespace QuantConnectStubsGenerator.Parser
         {
             var intImplicitOperatorMethod = new Method("__int__", new PythonType("int")) { Class = _currentClass };
             _currentClass.Methods.Add(intImplicitOperatorMethod);
+        }
+
+        /// <summary>
+        /// Rewrites typing.Iterable[T] to typing.Sequence[T] for a method return type.
+        /// Recurses into type parameters so nested forms like typing.Optional[typing.Iterable[T]]
+        /// are also rewritten. Safe to mutate in place: TypeConverter hands back fresh
+        /// PythonType instances for everything except the parameter-less PythonType.Any/None
+        /// singletons, which have no type parameters and are short-circuited.
+        /// </summary>
+        private static PythonType IterableReturnToSequence(PythonType type)
+        {
+            if (type.Namespace == "typing" && type.Name == "Iterable" && type.TypeParameters.Count == 1)
+            {
+                return new PythonType("Sequence", "typing")
+                {
+                    TypeParameters = { IterableReturnToSequence(type.TypeParameters[0]) }
+                };
+            }
+
+            for (var i = 0; i < type.TypeParameters.Count; i++)
+            {
+                type.TypeParameters[i] = IterableReturnToSequence(type.TypeParameters[i]);
+            }
+
+            return type;
         }
     }
 }

--- a/QuantConnectStubsGenerator/Parser/MethodParser.cs
+++ b/QuantConnectStubsGenerator/Parser/MethodParser.cs
@@ -270,6 +270,16 @@ namespace QuantConnectStubsGenerator.Parser
                     continue;
                 }
 
+                // Detect PyObject parameters from the original syntax — by the time we read
+                // parsedParameter.Type it has already been collapsed to typing.Any, which also
+                // comes from System.Object and other sources. Relying on the raw syntax lets
+                // us flag Python-equivalent overloads regardless of whether they live in a
+                // .Python.cs file or a plain .cs partial.
+                if (parameter.Type is SimpleNameSyntax simpleName && simpleName.Identifier.Text == "PyObject")
+                {
+                    method.HasPyObjectParameter = true;
+                }
+
                 method.Parameters.Add(parsedParameter);
 
                 if (parameter.Modifiers.Any(m => m.Text == "out"))
@@ -316,15 +326,15 @@ namespace QuantConnectStubsGenerator.Parser
             method.GenericType = genericType;
             method.AvoidImplicitTypes = avoidImplicitConversionTypes;
 
-            // When partial classes define the same signature in both a .Python.cs file and the
-            // plain .cs file (e.g. QCAlgorithm.Link(PyObject) vs QCAlgorithm.Link(object), both
-            // normalized to `command: Any`), HashSet dedupe is first-wins and ignores File.
-            // If the .Python.cs variant is visited first, PostProcessClass later deletes it
-            // as an override with no C# counterpart. Drop the .Python.cs entry so the .cs
-            // version can take its place.
-            if (!method.File.EndsWith(".Python.cs")
+            // When partial classes define the same signature in both a PyObject-taking and a
+            // plain-C# variant (e.g. QCAlgorithm.Link(PyObject) vs QCAlgorithm.Link(object),
+            // both normalized to `command: Any`), HashSet dedupe is first-wins and ignores
+            // which is which. If the PyObject variant is visited first, PostProcessClass later
+            // deletes it as an override with no C# counterpart. Drop the PyObject entry so the
+            // C# version can take its place.
+            if (!method.HasPyObjectParameter
                 && _currentClass.Methods.TryGetValue(method, out var existing)
-                && existing.File.EndsWith(".Python.cs"))
+                && existing.HasPyObjectParameter)
             {
                 _currentClass.Methods.Remove(existing);
             }
@@ -384,27 +394,16 @@ namespace QuantConnectStubsGenerator.Parser
                 }
             }
 
-            // Methods like AddData<T> and History<T> have Python implementations accepting "T" as first parameter
-            // We set the types of these parameters to typing.Type instead of the default typing.Any
-            if (_model.SyntaxTree.FilePath.EndsWith(".Python.cs")
+            // Methods like AddData<T> and History<T> have Python wrappers that accept the data
+            // type as a PyObject first parameter (named "type"/"dataType"/"T"). Expose it as
+            // typing.Type instead of the default typing.Any. Detected via the syntax's raw
+            // PyObject declaration rather than a .Python.cs file-suffix check — PyObject-taking
+            // wrappers can live alongside their C# counterparts in plain .cs partials too.
+            if (syntax.Type is SimpleNameSyntax pyObjectSyntax
+                && pyObjectSyntax.Identifier.Text == "PyObject"
                 && (parameter.Name == "type" || parameter.Name == "dataType" || parameter.Name == "T"))
             {
                 parameter.Type = new PythonType("Type", "typing");
-            }
-
-            // PyObject tickers parameter: the Python.cs History wrapper accepts Symbol,
-            // List[Symbol], List[str], Universe, and Type (e.g. self.history(Fundamental, ...)),
-            // returning pandas.DataFrame in each case. Widen the typing so mypy picks this
-            // overload over the C# IEnumerable one.
-            if (parameter.Name == "tickers" && parameter.Type == PythonType.Any)
-            {
-                parameter.Type = PythonType.CreateUnion(
-                    new PythonType("Symbol", "QuantConnect"),
-                    new PythonType("List", "typing") { TypeParameters = { new PythonType("Symbol", "QuantConnect") } },
-                    new PythonType("List", "typing") { TypeParameters = { new PythonType("str") } },
-                    new PythonType("Universe", "QuantConnect.Data.UniverseSelection"),
-                    new PythonType("Type", "typing")
-                );
             }
 
             // System.Object parameters can accept anything

--- a/QuantConnectStubsGenerator/Parser/TypeConverter.cs
+++ b/QuantConnectStubsGenerator/Parser/TypeConverter.cs
@@ -286,6 +286,23 @@ namespace QuantConnectStubsGenerator.Parser
                     isList = true;
                 }
             }
+            else if (type.Namespace == "System.Collections.Generic" && type.TypeParameters.Count == 2)
+            {
+                if (type.Name == "IReadOnlyDictionary")
+                {
+                    // typing.Mapping is the covariant read-only counterpart: it accepts
+                    // dict[K, V] subtypes at assignment sites, which lets user algorithms
+                    // override properties like DefaultMarkets with a plain dict literal.
+                    return new PythonType("Mapping", "typing")
+                    {
+                        TypeParameters =
+                        {
+                            NormalizeType(type.TypeParameters[0], isParameter),
+                            NormalizeType(type.TypeParameters[1], isParameter)
+                        }
+                    };
+                }
+            }
             else if (type.Namespace == "System.Collections" && type.Name == "IList")
             {
                 isList = true;

--- a/QuantConnectStubsGenerator/Parser/TypeConverter.cs
+++ b/QuantConnectStubsGenerator/Parser/TypeConverter.cs
@@ -290,10 +290,12 @@ namespace QuantConnectStubsGenerator.Parser
             {
                 if (type.Name == "IReadOnlyDictionary")
                 {
-                    // typing.Mapping is the covariant read-only counterpart: it accepts
-                    // dict[K, V] subtypes at assignment sites, which lets user algorithms
-                    // override properties like DefaultMarkets with a plain dict literal.
-                    return new PythonType("Mapping", "typing")
+                    // Expose as typing.Dict — pythonnet presents IReadOnlyDictionary<K, V>
+                    // with full dict semantics at runtime, and dict is the idiom users already
+                    // know. Using Dict (invariant) rather than Mapping (covariant read-only)
+                    // keeps overrides like `default_markets = { ... }` accepted without
+                    // introducing a less familiar ABC.
+                    return new PythonType("Dict", "typing")
                     {
                         TypeParameters =
                         {

